### ISSUE TITLE
Texture Bug in Dissidia Final Fantasy on GPU PowerVR

### ISSUE THREAD

**Issue by zminhquanz** [hrydgard/ppsspp#7740]
Here this picture
![screenshot_2015-05-11-11-32-43 1](https://cloud.githubusercontent.com/assets/11285640/7591725/a161beb0-f8f9-11e4-8809-9adef0c0b769.png)

**Comment by asbel123** [comment]
are u using iphone on ppsspp

**Comment by zminhquanz** [comment]
@asbel123 i'm using Iphone and Zenfone using Intel SOC ( using PowerVR GPU )

**Comment by hrydgard** [comment]
What is the issue? Missing characters? It's not clear to me how it should look from the picture.. I don't play that game.

**Comment by zminhquanz** [comment]
@hrydgard yes , it's missing character when i switch Buffered Render , when i switch back to non-buffered render , it's normal . The issue is appear on GPU PowerVR, another GPU , Mali is not happen this

**Comment by asbel123** [comment]
just get samsung galaxy s6 edge u wont ppsspp problem i got lots of game fullspeed u can do 5x resoultion no slow down try final fantasy 7 crisi core is full speed i didnt put frameskip and try dissdia final fantasy it will be fullspeed.

powervr gpu is good its a problem is opengl version problems and iphone has lots of bugs and its no easy to hack new firmware now. andriod is best version to play ppsspp no problem.

**Comment by zminhquanz** [comment]
@asbel123 but , it's picture i'm screenshot is using ASUS Zenfone ( Intel using PowerVR GPU )

**Comment by eddiebrook** [comment]
@asbel123 
i dont have this problem on iphone 6. 
both OS are good but iOS is optimized and no need to much hardware like android.

Samsung S6 still can not match Nvidia Tegra k1 perfomance in dolphin emu. most games run slow and overheat alot because Mali diver are slow and  limit cpu.

**Comment by zminhquanz** [comment]
I think texture bug ( see picture on the top page ) is happen on Intel chip (I'm using ASUS Zenfone and Intel SOC using PowerVR ) @eddiebrook

**Comment by unknownbrackets** [comment]
It would be great if someone using a PowerVR GPU on an ARM device could confirm whether this is Intel specific (maybe some SSE3-only issue?) or PowerVR specific.

-[Unknown]

**Comment by zminhquanz** [comment]
![screenshot_2015-05-17-19-39-29](https://cloud.githubusercontent.com/assets/11285640/7670523/58b27ce0-fcd2-11e4-8653-6bd8b77cc704.png)
My Tablet is using GPU PowerVR SGX544 ( Mediatek MT8389T) like as ASUS Zenfone is also this bug like that

**Comment by zminhquanz** [comment]
I think excactly problem is on GPU PowerVR SGX544 , other GPU don't have this issue

**Comment by brujo5** [comment]
Work fine ON iPhone 5S PowerVR (serie 6)  G6430 

Using latest build

![image](https://cloud.githubusercontent.com/assets/4172898/7671132/e3b6e94e-fc7c-11e4-957b-8d1c16bb1417.jpg)

**Comment by zminhquanz** [comment]
@brujo5 you using non-buffered rendering or Buffered Rendering ?

**Comment by brujo5** [comment]
Bufered rendering 3x.

**Comment by zminhquanz** [comment]
@brujo5  so , i update the latest build , this bug isn't resolved and fixed , i think PowerVR on Android and IOS is different

**Comment by zminhquanz** [comment]
and i also think , PowerVR SGX544 ( Series 5 ) have this bug , but on Series 6 is isn't have this bug

**Comment by brujo5** [comment]
Or intel soc specific issue 

Ask for help on forum maybe someone have an diferent device/OS whit the same GPU PowerVR SGX544

**Comment by jccultima123** [comment]
I tried to hack the graphics config using GLTools 1.99 (XDA) by enabling stencil hacks, and it works. However, i got weird missing artifacts on each model (including environment) especially in BR mode.

For example: Cloud's buster sword still appears on front.. (sorry, screenshot not yet available)

**Comment by unknownbrackets** [comment]
Sounds like this is a bug specific to all PowerVR SGX 544 devices (not just x86.)  It's hard to tell exactly what's going wrong, and I'm not sure exactly what "stencil hacks" do.

The character is drawn with:

Texture L0: 0x08d6ed80 (256x256)
Stencil test - pass if (80 & 80) ALWAYS (a & 80)
Depth test - pass if src >= dst
Alpha test - pass if (a & ff) > (04 & ff)
Alpha blend mode - add: src.a, 1.0 - src.a
(the buffer is 5551.)

So, it shouldn't matter what stencil is.  The depth values are pretty far apart too.  The texture's alpha is usually full, so it shouldn't even be an alpha testing problem.

Bloom is also applied after drawing the character, with the character.

This also renders properly with stencil to alpha disabled.

-[Unknown]

**Comment by hrydgard** [comment]
If the driver is really silly, maybe it would help to set the stencil test mask to 0xFF if operation is ALWAYS? Otherwise I don't really know what to do about it..

**Comment by unknownbrackets** [comment]
Yeah, that's probably a good idea, #8779.

-[Unknown]

**Comment by zminhquanz** [comment]
it's may fix this issue missing texture in other game like Dissidia 012

**Comment by jccultima123** [comment]
Gonna test this tonight with device default configs, and then the GLTools hacks.. i'll give a comparison screenshot later

**Comment by jccultima123** [comment]
Here's what i got, i am using the latest build (v1.2.2-592-gda34e31):

PS: I am using Dissidia playable demo, since they have same issues on 012 duodecim version

But before that, here's my current GLTools config.. the `GL_STENCIL_INDEX8` hack was the thing on my PowerVR SGX 544 issue.. if it's activated, the game would be playable in BR but still missing artifacts
![screenshot_2016-05-29-20-46-12](https://cloud.githubusercontent.com/assets/4709030/15633422/fc59b972-25de-11e6-8d17-4c6f5c076748.jpg)

Now, Here's the result if `GL_STENCIL_INDEX8` was disabled. Even the overall preset disabled (device default values)
![screenshot_2016-05-29-20-44-06](https://cloud.githubusercontent.com/assets/4709030/15633444/61d26e02-25df-11e6-8718-0ac21d74f181.jpg)

Here's interesting when `GL_STENCIL_INDEX8` was activated
![screenshot_2016-05-29-20-47-50](https://cloud.githubusercontent.com/assets/4709030/15633455/b10dcdcc-25df-11e6-9a00-888173d4c0a8.jpg)
![screenshot_2016-05-29-20-47-54](https://cloud.githubusercontent.com/assets/4709030/15633454/b10c35d4-25df-11e6-8ece-c9052983e778.jpg)
![screenshot_2016-05-29-20-48-14](https://cloud.githubusercontent.com/assets/4709030/15633456/b10fb7fe-25df-11e6-908b-de4b92aa8470.jpg)
![screenshot_2016-05-29-20-48-23](https://cloud.githubusercontent.com/assets/4709030/15633457/b1361a98-25df-11e6-81d0-5527d44afacd.jpg)

**Comment by unknownbrackets** [comment]
Just to mention - GLTools basically lies to PPSSPP and the only thing you'll usually get out of it is worse performance.  We may have detection code to use `glMapBuffer` only when your device supports it natively and it's fast, and use other methods when it's not there.  If it lies and says you have it, we might use it even when it's the slow-mo way.

Just to confirm, this still happens with all GLTools hacks disabled, right?  I assume that's what you mean by "device default values".

I'm not sure what that setting does.  I suspect it's causing depth not to work since the device already natively supports packed depth/stencil buffers, e.g. making us not pick packed even when that's the fastest and best option for your device.

To confirm, can you look at an "adb logcat" with the settings that way?  If it says "Creating 480 x 272 FBO using separate stencil", then that's what's happening.

-[Unknown]

**Comment by jccultima123** [comment]
@unknownbrackets unfortunately, my device seems that it doesn't fetch main logs in my logcat

**Comment by zminhquanz** [comment]
@jccultima123 summary , if you don't using GLtools to have fix the issues that

**Comment by jccultima123** [comment]
Finally, i already fetched the logcat so i begin tests earlier.. here's what i got @unknownbrackets 

`I/NativeApp( 4939): Creating 480 x 272 FBO using DEPTH24_STENCIL8`

**Comment by zminhquanz** [comment]
@jccultima123 other a few game have a similarly bug , but i forget that name of this game

**Comment by jccultima123** [comment]
@zminhquanz Crisis Core FFVII has also this issue though

**Comment by zminhquanz** [comment]
@jccultima123 @unknownbrackets 
yeah i see the issue in Kingdom Heart birth by sleep USA , when i'm attack , combat and using skill , it's missing effect , switch back Non-buffered rendering , it's normal
![screenshot_2016-06-08-07-06-47](https://cloud.githubusercontent.com/assets/11285640/15878895/24eedcca-2d49-11e6-8a82-40d2cdb509b7.jpg)
![screenshot_2016-06-08-07-08-17](https://cloud.githubusercontent.com/assets/11285640/15878896/24f104c8-2d49-11e6-8145-a3ace238d435.jpg)

**Comment by jccultima123** [comment]
@zminhquanz yeah, that's also included in this issue. In fact, some Square Enix games was still unable to be playable at least in BR mode. They using extensive use of effects for these games

**Comment by unknownbrackets** [comment]
A few things we don't do in non-buffered rendering:

1. Blit depth, but that requires GLES3 anyway.
2. Reformat framebuffers, which Kingdom Hearts uses for shadows.
3. Clear the backbuffer color, depth, and stencil to zero before drawing a frame.
4. Upload initial stencil and memory values to framebuffers.

When skipping buffer effects, we do clear the backbuffer at the very start to have alpha = 1.  This is wrong but probably doesn't matter much since games typically clear anyway.

We've seen before that the PowerVR drivers have bad problems with blending and with alpha testing.  It seems like it could be anything here:

 * Maybe the `>=` depth test is too hard for it.
 * Maybe the alpha test against something other than zero is too hard for it, but operates differently when the framebuffer format is different.
 * Maybe stencil testing at the same time as alpha blending is too hard for it.
 * Maybe we are choosing a non-8888 buffer for the backbuffer, and the driver has less bugs with this format.  Perhaps the hack is affecting framebuffer selection.

Not sure.

-[Unknown]

**Comment by zminhquanz** [comment]
I don't think so , driver from ASUS is good , not bad from Chinese manufacturer

**Comment by jccultima123** [comment]
@zminhquanz let me clear about that. PowerVR drivers are built according to their target model/capabilities and OpenGL version. I believe the drivers have the same capabilities on all devices. 

Therefore, the issue is the PowerVR itself since they are providing lack of graphical features for low to middle-entry devices.

**Comment by unknownbrackets** [comment]
That's not really the case.  A lot of older phones don't have updated drivers, sometimes.  It's not as if video drivers are magically exempt from the rule that all software contains bugs.

For example, I remember there was a blending bug that affected most PowerVR SGX 5xx device drivers.  It's potentially an issue here too, but at least it's not affecting the character drawing in Dissidia.

All I know about the bug was that blending with GL_CONSTANT_COLOR / 0xFFFFFF as the blend factor, instead of GL_ONE (which is equivalent), didn't work.  Potentially this means other blend factors, like GL_CONSTANT_COLOR / 0xF0F0F0 (which is not the same as GL_ONE) also don't work.

This isn't a "new" OpenGL feature or one that has optional support or is only required by certain versions.  It's supposed to work.  It works everywhere else.  At least for 0xFFFFFF, the driver could've made it work, and didn't.  The hardware was capable of at least that value, for sure.

We have other workarounds in the code for known bugs in PowerVR, including the modulus operator simply not operating according to the specification - we have to implement our own.  Again, not an optional thing.  Not a hardware deficiency.  It's just wrong.

FWIW the Rogue (G6xxx) PowerVR devices seem to be better in this regard.  At least it seems to have a lower density of bugs in its drivers, based on reports.

-[Unknown]

**Comment by zminhquanz** [comment]
Maybe you can implement or rewrite a code

**Comment by jccultima123** [comment]
@unknownbrackets  I see. In my case before, i start to test games since the 0.7.5 release. From there, it was relevant to this issue before (even my drivers was OpenGL 3 capable) until it's fixed in 0.9

So it's clear that these drivers including mine has potential bugs or maybe not capable. I don't know

**Comment by zminhquanz** [comment]
#1313

**Comment by afifah1978** [comment]
This file is optional
# for syntax explanation check:
# - https://github.com/hrydgard/ppsspp/pull/8715 
# - https://github.com/hrydgard/ppsspp/pull/8792

**Comment by unknownbrackets** [comment]
Is this still an issue?  There have been lots of changes that might have improved this since.  These devices are also becoming less commonly used...

-[Unknown]

**Comment by Gamemulatorer** [comment]
This is not a dup of https://github.com/hrydgard/ppsspp/issues/9951?

**Comment by Gamemulatorer** [comment]
Anyone with powervr gpu can still reproduce this issue using v1.14???

**Comment by Gamemulatorer** [comment]
@zminhquanz no feedback outdated?

**Comment by Gamemulator00122719** [comment]
Any updated framedump that I can test on my redmi 6a?
